src/server/pkg/storage/chunk/reader.go

```diff
@@ -20,15 +25,18 @@ type Reader struct {
 	buf      *bytes.Buffer
 	r        *bytes.Reader
 	len      int64
+	f        ReaderFunc
 }
 
-func newReader(ctx context.Context, objC obj.Client, dataRefs ...*DataRef) *Reader {
+func newReader(ctx context.Context, objC obj.Client, f ...ReaderFunc) *Reader {
 	r := &Reader{
 		ctx:  ctx,
 		objC: objC,
 		buf:  &bytes.Buffer{},
 	}
-	r.NextRange(dataRefs)
+	if len(f) > 0 {
+		r.f = f[0]
+	}
```

Comment:
So if I'm understanding this correctly `f` is variadic so as to make it optional, there should never be more than 1 passed right? What's the point of this struct if `f` is `nil`, is it just an empty reader? If so could that just be represented by a function that returns nothing, rather than a `nil` function. I think that might be a little bit clearer.

Reply:
Yes to the reason it is variadic. I'm not sure I follow what you are saying after the first question. If they do not pass in a reader function, then read calls will return io.EOF until they setup a new range (NextRange). The basic pattern is to setup a range then make read calls until io.EOF (new ranges can be setup multiple times). The optional reader function is used for streaming bytes (the reader can request more data refs through the callback when io.EOF is reached for the current range). The usage is, in general, either the client making multiple NextRange calls or the reader using the callback to stream bytes.

Reply:
Ahh, I see so there's actually 2 different ways that dataRefs get into the reader, they can be pushed in with `NextRange` or pulled in by calls to `f`. That makes sense and clarifies my comment below as well.